### PR TITLE
Removing unnecessary iterator operations from op_apply

### PR DESCRIPTION
**Pull request by VitalyFedyunin**
In many cases we are iterating tensors with one continuous memory (any stride size). That means there is no need to check overflow/increase counters. It saves 3 asm operations per tensor (out of 13 in copy case).

Simple test:

```
import torch
torch.set_num_threads(1)
i = torch.arange(100000000).type(torch.DoubleTensor).reshape(20,500,100,100)
timeit i.transpose(0,2).transpose(1,2).clone()
```

With optimization:
962 ms ± 17.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

Without:
1.03 s ± 22.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

There is still room to apply it in all CPU_tensor_* functions. I just covered most common ones.

### CODE REVIEW

**Comment by ezyang** [comment]
cc @cpuhrsch @colesbury @umanwizard

File: aten/src/ATen/CPUApplyUtils.h [hunk 1]
```diff
@@ -151,6 +151,7 @@ struct strided_tensor_iter_fixed {
   int64_t strides_[N] = {0};
 
   strided_tensor_iter_fixed(strided_tensor_iter_fixed const&) = delete;
+  operator bool() const { return CanOverflow; };
```
Comment (ezyang) [inline aten/src/ATen/CPUApplyUtils.h]: I found this a bit unexpected. Why does overflow have anything to do with whether or not an iterator converts to bool? Why should "can overflow" be privileged in this way?

Comment (VitalyFedyunin) [reply aten/src/ATen/CPUApplyUtils.h]: Agree, changing to:
bool Overflow() const { return CanOverflow; };

**Comment by VitalyFedyunin** [comment]
note: much more easier to read if you hide whitespace changes in github diff

File: aten/src/ATen/CPUApplyUtils.h [hunk 2]
```diff
@@ -151,6 +151,7 @@ struct strided_tensor_iter_fixed {
   int64_t strides_[N] = {0};
 
   strided_tensor_iter_fixed(strided_tensor_iter_fixed const&) = delete;
+  bool Overflow() const { return CanOverflow; };
```
Comment (ezyang) [inline aten/src/ATen/CPUApplyUtils.h]: This is the nittiest of nits, but could we please use a method name that matches the prevailing method naming style (namely, lower case method)?

Comment (ezyang) [reply aten/src/ATen/CPUApplyUtils.h]: While we're on the topic, let's have a comment? :)

**Comment by ezyang** [comment]
I found this diff somewhat hard to understand, I think part of it is due to some funny business in the current code structure. It might be right to take this patch as is, but I think this patch could be clearer if we did a little refactoring first.

Right now, we first test what the dimensionality of the tensors are (to decide if we shunt into the optimized codepath) and the collapse the dimensions. This seems wrong: we should *first* collapse dimensions and then decide if we can go in the codepath. There are multiple benefits of this:

1. More tensors will be eligible for the optimized codepath. If I have a 100-dim contiguous tensor, I don't get to use the fixed iterator. There's no reason why I shouldn't be able to use it in this case.
2. The condition for `CanOverflow` is simplified. Now it is equivalent to `N = 1`. So we can eliminate the extra parameter entirely. (You could say `N = 8` in that case, but that's just wasting runtime space.)

One extra thing to note: with this change, we generate an *extra* copy of the kernel, specifically specialized for the contiguous case. This seems reasonable to me, but it's something to be aware of from a binary size perspective. Might be worth measuring what binary size change was.

@cpuhrsch, could you also review this diff (and my comment?)

**Comment by VitalyFedyunin** [comment]
Btw, I've removed dimensionality test to make sure that we are not slowing down existing functionality. Totally agree about 100dim contiguous tensor thou. I will handle such cases in next PR on top of it. 

Numbers of kernels is a valid concern if we plan to apply same patter to all CPU_ops. As for now it is only 2 additional kernels per op to cover the most popular use-case (clone() function).

**Comment by ezyang** [comment]
@fmassa wanted to know if we should just use TensorIterator directly instead.

**Comment by cpuhrsch** [comment]
In general, TensorIterator has supplanted a lot of the code in this file. @colesbury has the necessary context here.

We should get a list of callsites of these functions to see what is affected and whether it's possible replacing it with TensorIterator.

The difference in timing here is pretty small, so it's hard to say how precise it is. It's also restricted to a single function. When building this code we looked at various tensor shapes and in particular unary operations because those were the first customers. This requires precise control of the environment. Vitaly and I are talking about this right now.

cc @apaszke who worked on this with me.

**Comment by cpuhrsch** [comment]
As in aside, it might be worth looking into removing CPUApplyUtils.h entirely if it's been replaced by TensorIterator. Then this PR's insight can be used to improve TensorIterator and we further reduce the complexity of our codebase.

File: aten/src/ATen/CPUApplyUtils.h [hunk 3]
```diff
@@ -538,18 +556,33 @@ inline void CPU_tensor_parallel_apply2(
...
+      parallel_for(
+          0,
+          tensor1.numel(),
+          grain_size,
+          [&tensor1, &tensor2, &op](int64_t begin, int64_t end) {
+            apply_op(
+                end - begin,
+                begin,
+                op,
+                strided_tensor_iter_fixed<scalar1, 8, false>(tensor1),
```
Comment (apaszke) [inline aten/src/ATen/CPUApplyUtils.h]: Looks like you haven't checked that `tensor2` is contiguous?

Comment (VitalyFedyunin) [reply aten/src/ATen/CPUApplyUtils.h]: Purposely as this apply operation primarily used in copy (clone) function where destination is contiguous and source not (as when they are both contiguous we do memcpy).

Comment (fmassa) [reply aten/src/ATen/CPUApplyUtils.h]: Does that mean that `t1.copy_(t2)`, for t1 non-contiguous (or t2, didn't check the rest of the code) will give wrong results?

Comment (VitalyFedyunin) [reply aten/src/ATen/CPUApplyUtils.h]: It is only mean that it will use older non-optimized kernel. I can do kernel per each couple: (non-contiguous, non-contiguous).(contiguous,non-contiguous),... but it blows up exponentially for cpu_apply that takes 3 or more tensors.
